loopy/codegen/tools.py

```diff
@@ -0,0 +1,152 @@
+__copyright__ = "Copyright (C) 2020 Kaushik Kulkarni"
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+
+from pytools import ImmutableRecord, memoize_method
+from loopy.schedule import (EnterLoop, LeaveLoop, CallKernel, ReturnFromKernel,
+                            Barrier, BeginBlockItem, gather_schedule_block)
+
+
+__doc__ = """
+.. currentmodule:: loopy.codegen.tools
+
+.. autoclass:: CodegenOperationCacheManager
+"""
+
+
+class CodegenOperationCacheManager(ImmutableRecord):
```

Comment:
Somehow, it seems to me as if the schedule (linearization) should be an object and this should be attached to that? Or maybe this should be under `loopy.schedule` and be renamed to something like `LinearizationOperationCacheManager`?

Reply:
These operations are invoked during the codegen pipeline, so I went with `CodegenOperationCacheManager`. It also uses more information than the LoopKernel's linearization such as the iname_tags, so it cannot be tied to the linearization obect IMO.